proxy/api/src/project.rs

```diff
@@ -69,6 +73,139 @@ where
+    /// # Errors
+    ///
+    ///   * We couldn't get the list of projects
+    ///   * We couldn't inspect the `signed_refs` of the project
+    ///   * We couldn't get stats for a project
+    pub fn list(state: &coco::State) -> Result<Self, error::Error> {
+        let mut projects = Self {
+            tracked: vec![],
+            contributed: vec![],
+        };
+        for project in state.list_projects()? {
+            let refs = state.list_owner_project_refs(&project.urn())?;
+            let project = state.with_browser(&project.urn(), |browser| {
+                let project_stats = browser.get_stats().map_err(coco::Error::from)?;
+                Ok((project, project_stats).into())
+            })?;
+            if refs.heads.is_empty() {
+                projects.tracked.push(Tracked(project))
+            } else {
+                projects.contributed.push(project)
+            }
+        }
+
+        Ok(projects)
+    }
+
+    /// Give back an `Iter` that can be used to iterate over the projects. It first yields
+    /// contributed projects and then tracked projects.
+    pub fn iter(&self) -> Iter<'_> {
+        Iter {
+            contributed: self.contributed.iter(),
+            tracked: self.tracked.iter(),
+        }
+    }
+}
+
+/// An iterator over [`Projects`] that first yields contributed projects and then tracked projects.
+#[must_use = "iterators are lazy and do nothing unless consumed"]
+pub struct Iter<'a> {
+    /// Iterator over contributed projects.
+    contributed: std::slice::Iter<'a, Project>,
+
+    /// Iterator over tracked projects.
+    tracked: std::slice::Iter<'a, Tracked>,
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = &'a Project;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.contributed
+            .next()
+            .or_else(|| match self.tracked.next() {
+                Some(tracked) => Some(&tracked.0),
+                None => None,
+            })
+    }
+}
+
+impl IntoIterator for Projects {
+    type Item = Project;
+    type IntoIter = IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        IntoIter {
+            contributed: self.contributed.into_iter(),
+            tracked: self.tracked.into_iter(),
+        }
+    }
+}
+
+/// An iterator over [`Projects`] that moves the values into the iterator.
+/// It first yields contributed projects and then tracked projects.
+#[must_use = "iterators are lazy and do nothing unless consumed"]
+pub struct IntoIter {
+    /// Iterator over contributed projects.
+    contributed: std::vec::IntoIter<Project>,
+
+    /// Iterator over tracked projects.
+    tracked: std::vec::IntoIter<Tracked>,
+}
+
+impl Iterator for IntoIter {
+    type Item = Project;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.contributed
+            .next()
+            .or_else(|| match self.tracked.next() {
+                Some(tracked) => Some(tracked.0),
+                None => None,
+            })
+    }
+}
+
```

Comment:
hm, am I getting it right that when we iterate on a `Projects` that it will return an iterator on the `contributed` projects implicitly?

Reply:
It returns an iterator over _all_ the projects whether tracked or contributed

Reply:
Ah I follow now. So we first iterate over the `contributed` and when that's done we iterate over the `tracked`. Thanks!

Reply:
Exactly :white_check_mark: